features/base/conference/middleware.js

```diff
@@ -0,0 +1,79 @@
+import JitsiMeetJS from '../lib-jitsi-meet';
+import { MiddlewareRegistry } from '../redux';
+import {
+    TRACK_ADDED,
+    TRACK_REMOVED
+} from '../tracks';
+
+const JitsiTrackErrors = JitsiMeetJS.errors.track;
+
+/**
+ * This middleware intercepts TRACK_ADDED and TRACK_REMOVED actions to sync
+ * conference's local tracks with local tracks in state.
+ *
+ * @param {Store} store - Redux store.
+ * @returns {Function}
+ */
+const conferenceMiddleware = store => next => action => {
+    let actionType = action.type;
+    let track = action.track;
+
+    if ((actionType === TRACK_ADDED || actionType === TRACK_REMOVED) &&
+        track &&
+        track.isLocal()) {
+        return syncConferenceLocalTracksWithState(store, action)
+            .then(() => next(action));
+    } else {
+        return next(action);
+    }
+};
+
+/**
+ * Syncs local tracks from state with local tracks in JitsiConference instance.
+ *
+ * @param {Store} store - Redux store.
+ * @param {Object} action - Action object.
+ * @returns {Promise}
+ */
+const syncConferenceLocalTracksWithState = (store, action) => {
+    const conference = store.getState()['features/base/conference'];
+    let promise = Promise.resolve();
+
+    if (conference) {
+        let actionType = action.type;
+        let track = action.track;
+
+        if (actionType === TRACK_ADDED) {
+            const conferenceLocalTracks = conference.getLocalTracks();
+
+            // XXX The library lib-jitsi-meet may be draconian, for example,
+            // when adding one and the same video track multiple times.
+            if (conferenceLocalTracks.indexOf(track) === -1) {
+                promise = conference.addTrack(track)
+                    .catch(err => {
+                        // TODO: This is a good point to call some global
+                        // error handler when we have one.
+                        console.error(
+                            'Failed to add local track to conference', err);
+                    });
+            }
+        } else {
+            promise = conference.removeTrack(track)
+                .catch(err => {
+                    // Local track might be already disposed by direct
+                    // JitsiTrack#dispose() call. So we should ignore this
+                    // error here.
+                    if (err.name !== JitsiTrackErrors.TRACK_IS_DISPOSED) {
+                        // TODO: This is a good point to call some global
+                        // error handler when we have one.
+                        console.error('Failed to remove local track to ' +
+                            'conference', err);
```

Comment:
As I just commented, please don't modify this PR because I've started its review & merge.

I'd like to point out this example though. As is made obvious by the duplicated TODO comments, a problem has been identified. The identifications & its marking with TODO are the right steps in my mind.

What I think is missing is the containment of the problem. I would've expected these 2 TODO + console.error to be extracted into a suitably named function which reduces the 2 instances of the problem to 1.